internal/beater/config/profiling.go

```diff
@@ -0,0 +1,89 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package config
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// ProfilingConfig holds configuration related to profiling.
+type ProfilingConfig struct {
+	Enabled bool `config:"enabled"`
+
+	ESConfig *elasticsearch.Config
+
+	es *config.C
+}
+
+func (c *ProfilingConfig) Unpack(in *config.C) error {
+	type profilingConfig ProfilingConfig
+	cfg := profilingConfig(defaultProfilingConfig())
+	if err := in.Unpack(&cfg); err != nil {
+		return errors.Wrap(err, "error unpacking config")
+	}
+	cfg.Enabled = in.Enabled()
+	*c = ProfilingConfig(cfg)
+
+	if in.HasField("elasticsearch") {
+		es, err := in.Child("elasticsearch", -1)
+		if err != nil {
+			return err
+		}
+		c.es = es
+	}
+
+	return errors.Wrap(c.Validate(), "invalid config")
+}
+
+func (c *ProfilingConfig) Validate() error {
+	return nil
+}
+
+func (c *ProfilingConfig) setup(log *logp.Logger, outputESCfg *config.C) error {
+	if !c.Enabled {
+		return nil
+	}
+	// NOTE(axw) this behaviour below is a bit different to what we have
+	// in other code that has Elasticsearch config. Here we always merge
+	// the provided Elasticsearch config over the top of the Elasticsearch
+	// output config. This allows one to provide only an API Key for
+	// profiling (-E apm-server.profiling.elasticsearch.api_key=...), but
+	// still use the same Elasticsearch output host, etc.
```

Comment:
nit: are there use cases for configuring something other than the API Key at this point? If not, should we only extract the API Key for now to avoid weird edge cases where other ES configs (URLs, credentials, etc) get weirdly merged with each other?

Reply:
I'm inclined to leave it, to allow some flexibility for 8.5. It'll get ripped out soon enough.